### PR TITLE
[Model] Heterogeneous graph support for GNNExplainer

### PR DESCRIPTION
**Pull request by moeinfinityx**
## Description

Add Heterogeneous Graph support for GNNExplainer

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes

Added a new class HeteroGNNExplainer for Heterogeneous graph explanation.

The output of the model has been manually verified on MUTAG dataset. Unfortunately there's no ground truth for the explanation, so we cannot quantify the model's performance. Instead, we manually investigated the results and confirms that it aligns well with human intuition.

[Link to the MUTAG Hetero GNNExplainer notebook](https://github.com/ahadjawaid/Syssec/blob/main/data_converter/MUTAGHetro_Graph_Explainer.ipynb)

The code has also been verified on several datasets DGL officially provides, see [this repo](https://github.com/ahadjawaid/Syssec/tree/main/Explainer_Notebooks) for details.

Inline documentation is added following the original style, and an example of using the module is included in the documentation.

### CODE REVIEW

File: python/dgl/nn/pytorch/explain/gnnexplainer.py [hunk 1]
```diff
@@ -425,3 +398,394 @@ def explain_graph(self, graph, feat, **kwargs):
...
+            the more important.
+        """
+        self.model.eval()
+
+        # Get the initial prediction.
+        with torch.no_grad():
+            logits = self.model(graph, feat, **kwargs)
+            pred_label = logits.argmax(dim=-1)
+
+        feat_mask, edge_mask = self._init_masks(graph, feat)
```
Comment (jikk) [inline python/dgl/nn/pytorch/explain/gnnexplainer.py:754]: why don't we name variables differently? Say  `feat_mask` --> `feat_masks` or `feat_mask_map` 
AFAIK, the structural differences of `feat_mask` and `edge_mask` are among our key considerations.

**Review by jikk — COMMENTED** [review]
minor comments

File: python/dgl/nn/pytorch/explain/gnnexplainer.py [hunk 2]
```diff
@@ -425,3 +425,269 @@ def explain_graph(self, graph, feat, **kwargs):
         edge_mask = edge_mask.detach().sigmoid()
 
         return feat_mask, edge_mask
+
+
+class HeteroGNNExplainer(nn.Module):
+    coeffs = {
```
Comment (muhyun) [inline python/dgl/nn/pytorch/explain/gnnexplainer.py]: Need to put the reference where these values are from.

File: python/dgl/nn/pytorch/explain/gnnexplainer.py [hunk 3]
```diff
@@ -425,3 +425,269 @@ def explain_graph(self, graph, feat, **kwargs):
...
+        'edge_size': 0.005,
+        'edge_ent': 1.0,
+        'node_feat_size': 1.0,
+        'node_feat_ent': 0.1
+    }
+
+    def __init__(self, model, num_hops, lr=0.01, num_epochs=100):
```
Comment (muhyun) [inline python/dgl/nn/pytorch/explain/gnnexplainer.py]: This does not have `log` parameter which controls the log output, which is defined in DGL's GNNExplainer.

**Comment by muhyun** [comment]
How about adding comments into https://github.com/ahadjawaid/Syssec/blob/main/data_converter/MUTAGHetro_Graph_Explainer.ipynb so that it could be used as both an example and a kickstart guide. 

https://github.com/ahadjawaid/Syssec/tree/main/Explainer_Notebooks. Let's have readme.md file to explain the notebooks .

**Comment by jikk** [comment]
> How about adding comments into https://github.com/ahadjawaid/Syssec/blob/main/data_converter/MUTAGHetro_Graph_Explainer.ipynb so that it could be used as both an example and a kickstart guide.
> 
> https://github.com/ahadjawaid/Syssec/tree/main/Explainer_Notebooks. Let's have readme.md file to explain the notebooks .

Agreed.

File: python/dgl/nn/pytorch/explain/gnnexplainer.py [hunk 4]
```diff
@@ -425,3 +398,394 @@ def explain_graph(self, graph, feat, **kwargs):
         edge_mask = edge_mask.detach().sigmoid()
 
         return feat_mask, edge_mask
+
+
+class HeteroGNNExplainer(nn.Module):
```
[the remark this answers is not in the document]
Comment (moeinfinityx) [reply python/dgl/nn/pytorch/explain/gnnexplainer.py:403]: Nothing is reused and the HeteroGNNExplainer will not accept homogeneous graph so maybe inherit from GNNExplainer won't help much.

File: python/dgl/nn/pytorch/explain/gnnexplainer.py [hunk 5]
```diff
@@ -12,32 +12,25 @@
 class GNNExplainer(nn.Module):
     r"""GNNExplainer model from `GNNExplainer: Generating Explanations for
     Graph Neural Networks <https://arxiv.org/abs/1903.03894>`__
-
```
[the remark this answers is not in the document]
Comment (moeinfinityx) [reply python/dgl/nn/pytorch/explain/gnnexplainer.py:15]: It's modified by the linter.

File: python/dgl/nn/pytorch/explain/gnnexplainer.py (hunk 1 above)
Comment (moeinfinityx) [reply python/dgl/nn/pytorch/explain/gnnexplainer.py:755]: Renamed to `feat_masks` and `edge_masks` as you suggested.

File: python/dgl/nn/pytorch/explain/gnnexplainer.py [hunk 6]
```diff
@@ -425,3 +398,394 @@ def explain_graph(self, graph, feat, **kwargs):
...
+        num_edges = graph.num_edges()
+
+        # Extract node-centered k-hop subgraph and
+        # its associated node and edge features.
+        sg, inverse_indices = khop_in_subgraph(graph, {ntype: node_id},self.num_hops)
+        inverse_indices = inverse_indices[ntype]
+        sg_nodes = sg.ndata[NID]
+        sg_edges = sg.edata[EID]
+        for node_type in feat.keys():
+            feat[node_type] = feat[node_type][sg_nodes[node_type]]
```
[the remark this answers is not in the document]
Comment (moeinfinityx) [reply python/dgl/nn/pytorch/explain/gnnexplainer.py]: Created a new variable `feat_sg` for the features of the subgraph.

File: python/dgl/nn/pytorch/explain/gnnexplainer.py (hunk 2 above)
Comment (moeinfinityx) [reply python/dgl/nn/pytorch/explain/gnnexplainer.py]: Already made them arguments, please see the latest comments.

File: python/dgl/nn/pytorch/explain/gnnexplainer.py (hunk 3 above)
Comment (moeinfinityx) [reply python/dgl/nn/pytorch/explain/gnnexplainer.py]: Already added it, please see the latest comments.

**Pull request merged**